src/lib/dateLib.ts

```diff
@@ -173,3 +178,206 @@ export const dateLib = {
+    this.dateLibBase.isBefore(...args);
+
+  /** Checks if the given value is a date. */
+  isDate: DateLibBase["isDate"] = (value): value is Date =>
+    this.dateLibBase.isDate(value);
+
+  /** Checks if the given dates are the same day. */
+  isSameDay: DateLibBase["isSameDay"] = (...args) =>
+    this.dateLibBase.isSameDay(...args);
+
+  /** Checks if the given dates are in the same month. */
+  isSameMonth: DateLibBase["isSameMonth"] = (...args) =>
+    this.dateLibBase.isSameMonth(...args);
+
+  /** Checks if the given dates are in the same year. */
+  isSameYear: DateLibBase["isSameYear"] = (...args) =>
+    this.dateLibBase.isSameYear(...args);
+
+  /** Returns the maximum of the given dates. */
+  max: DateLibBase["max"] = (...args) => this.dateLibBase.max(...args);
+
+  /** Returns the minimum of the given dates. */
+  min: DateLibBase["min"] = (...args) => this.dateLibBase.min(...args);
+
+  /** Sets the month for the given date. */
+  setMonth: DateLibBase["setMonth"] = (...args) =>
+    this.dateLibBase.setMonth(...args);
+
+  /** Sets the year for the given date. */
+  setYear: DateLibBase["setYear"] = (...args) =>
+    this.dateLibBase.setYear(...args);
+
+  /** Returns the start of the day for the given date. */
+  startOfDay: DateLibBase["startOfDay"] = (...args) =>
+    this.dateLibBase.startOfDay(...args);
+
+  /** Returns the start of an ISO week for the given date. */
+  startOfISOWeek: DateLibBase["startOfISOWeek"] = (...args) =>
+    this.dateLibBase.startOfISOWeek(...args);
+
+  /** Returns the start of the month for the given date. */
+  startOfMonth: DateLibBase["startOfMonth"] = (...args) =>
+    this.dateLibBase.startOfMonth(...args);
+
+  /** Returns the start of the year for the given date. */
+  startOfYear: DateLibBase["startOfYear"] = (...args) =>
+    this.dateLibBase.startOfYear(...args);
+
+  // ================================================
+  // DateLib functions which require locale overrides
+  // ================================================
+
+  /** Returns the end of the week for the given date. */
+  endOfWeek = <DateType extends Date>(date: DateArg<DateType>) => {
+    if (!this.locale) {
+      throw new Error(
+        "DateLib needs to be initialized with locale to use the `endOfWeek` function."
+      );
+    }
+
+    return endOfWeek(date, this.options);
+  };
+
+  /** Formats the given date using the specified format string. */
+  format = (date: DateArg<Date> & {}, formatStr: string) => {
+    if (!this.locale) {
+      throw new Error(
+        "DateLib needs to be initialized with locale to use the `format` function."
+      );
+    }
+
+    return format(date, formatStr, this.options);
+  };
+
+  /** Returns the week number for the given date. */
+  getWeek = (date: DateArg<Date> & {}) => {
+    if (!this.locale) {
+      throw new Error(
+        "DateLib needs to be initialized with locale to use the `getWeek` function."
+      );
+    }
+
```

Comment:
Also, can’t we rely on a default `this.options.locale` to avoid these checks? 🤔

Reply:
Default locale is what caused the original bug, by making locale optional on deeper calls in the library, we're allowing callers to forget to provide the known locale. For this implementation, I took the approach of still allowing locale optionally, but ensuring it's provided to use any function which is dependent on locale. This should help catch issues where locale is forgotten about in DateLib construction, but required when rendering.

Ultimately DateLib construction only happens inside `react-day-picker`, so the only time this is hit will be if we've written a bug, it's a footgun prevention mechanism